### PR TITLE
Merge remote-tracking branch 'dotnet/master' into NullableFeature

### PR DESCRIPTION
**Pull request by tannergooding**
CC. @safern, @krwq, @buyaa-n, @stephentoub 

This is primarily for bringing in the new compiler toolset.

### CODE REVIEW

**Comment by safern** [comment]
Should we instead rebase the branch? Will the merge commit show when we submit a PR to merge in master?

**Comment by tannergooding** [comment]
> Will the merge commit show when we submit a PR to merge in master?

Ultimately, yes. But that is also good as it shows how the branches were integrated together. This is just normal git workflow and is what almost every other repo in existence does.

For a closer to home comparison; Roslyn has multiple feature branches (which is what this branch effectively is) and has infrastructure setup to automatically flow the changes through between branches (release branches flow into master and master flows into feature branches).

**Comment by safern** [comment]
@krwq it looks like the new compiler is complaining about the workaround for field initialization:

`shared\System\Globalization\StringInfo.cs(29,13): error CS1717: Assignment made to same variable; did you mean to assign something else? [D:\a\1\s\src\System.Private.CoreLib\System.Private.CoreLib.csproj]`

https://github.com/dotnet/coreclr/blob/NullableFeature/src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs#L29

I guess we should disable this warning as well?

**Comment by tannergooding** [comment]
It should also be noted: **DO NOT** `Squash and merge` this or `Rebase and merge` this PR.

This PR must be merged with `Create a merge commit` for everything to work cleanly and to not result in further merge conflicts later down the line.

File: src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs [hunk 1]
```diff
@@ -24,9 +24,12 @@ public StringInfo() : this(string.Empty)
 
         public StringInfo(string value)
         {
-            this.String = value;
-            // TODO-NULLABLE: compiler doesn't see this field being initialized through property
-            _str = _str!;
+            // TODO-NULLABLE: This used to be: `this.String = value;`
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs]: @safern, @krwq. I worked around the issue by just manually inlining the initialization as I don't think disabling the self-assignment warning is the right way to workaround this.

CC. @jaredpar. @cston. Do we have a good story to allow bypassing this particular type of error when the user knows that the field was initialized to non-null?

File: src/System.Private.CoreLib/shared/Internal/Runtime/Augments/RuntimeThread.cs [hunk 2]
```diff
@@ -1,56 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (krwq) [inline src/System.Private.CoreLib/shared/Internal/Runtime/Augments/RuntimeThread.cs]: this remove feels out of place - was this intentional?

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/Internal/Runtime/Augments/RuntimeThread.cs]: It was removed from master and is just propagating the change back.

Comment (krwq) [reply src/System.Private.CoreLib/shared/Internal/Runtime/Augments/RuntimeThread.cs]: just noticed as well after seeing several more files

File: src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs (hunk 1 above)
Comment (cston) [reply src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs]: To work around the uninitialized non-nullable field warning:
disable the warning (`#pragma warning disable 8618`), or use `null!` at the declaration (`private string _str = null!;`). 

---
In reply to: [270507076](https://github.com/dotnet/coreclr/pull/23560#discussion_r270507076) [](ancestors = 270507076)

**Comment by krwq** [comment]
@tannergooding since this is not fixing any of the issues we have found and breaks workarounds, does it make sense to merge this right now? Perhaps we should wait for more fixes or at least better way to workaround the issues - I feel like in-lining code is more error-prone than the current workaround

File: src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs]: Thanks! Updated to do `_str = null!`

**Comment by tannergooding** [comment]
@krwq, yes it makes sense. This brings in the new compiler bits which includes the updated analysis rules and which will allow fixing several of the workarounds that have been made.

The actualy workarounds should be removed independently, however, and this is just touching up a single place where the build would otherwise fail. I've also updated this to no longer inline and to explicitly do `_str = null!` instead, which was indicated to be the correct resolution for these cases.

File: src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs (hunk 1 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs]: using `null!` looks like a good suggestion, @krwq could we do that in all your other changes? 😄

**Comment by safern** [comment]
Once this PR is merged, I'll run the build that creates the PR on master :smile:

File: src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs [hunk 3]
```diff
@@ -24,9 +24,8 @@ public StringInfo() : this(string.Empty)
 
         public StringInfo(string value)
         {
+            _str = null!;
```
Comment (krwq) [inline src/System.Private.CoreLib/shared/System/Globalization/StringInfo.cs]: actually, not too bad and easier to grep for

**Pull request merged**